lib/tree.rb

```diff
@@ -2,65 +2,339 @@ class TreeNode
+      return nil
+    else
+      if key == currNode.key
+        return currNode.value
+      elsif  key < currNode.key
+        return find_recursion_helper(currNode.left, key)
+      else
+        return find_recursion_helper(currNode.right, key)
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n) assuming a balanced tree, else O(n)
+  def inorder 
+    return inorder_helper(root, [])
+  end
+
+  def inorder_helper(currNode, list)
+
+    if currNode.nil?
+      # base case
+      return list
+
+    else
+      # recurse
+      # in order goes from left -> parent -> right
+      inorder_helper(currNode.left, list)
+    
+      list << { key: currNode.key, value: currNode.value }
+
+      inorder_helper(currNode.right, list)
+
+      return list
+    end
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n) assuming a balanced tree, else O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(currNode, list)
+
+    if currNode.nil?
+      # base case
+      return list
+
+    else
+      # recurse
+      # preorder goes from parent -> left -> right
+      list << { key: currNode.key, value: currNode.value }
+
+      preorder_helper(currNode.left, list)
+    
+      preorder_helper(currNode.right, list)
+
+      return list
+    end
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n) assuming a balanced tree, else O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(currNode, list)
+
+    if currNode.nil?
+      # base case
+      return list
+
+    else
+      # recurse
+      # postorder goes from left -> right -> parent
+
+      postorder_helper(currNode.left, list)
+    
+      postorder_helper(currNode.right, list)
+
+      list << { key: currNode.key, value: currNode.value }
+
+      return list
+    end
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n) assuming a balanced tree, else O(n)
   def height
-    raise NotImplementedError
+    return subtree_height(root)
+  end
+
+  def subtree_height(currNode)
+    if currNode.nil?
+      return 0
+    else
+      if !currNode.left && !currNode.right
+        return 1
+      else
+        # set initial conditions for the subtrees  
+        leftHeight = 0
+        rightHeight = 0
+
+        if currNode.left 
+          leftHeight = 1 + subtree_height(currNode.left)
+        end
+        
+        if currNode.right
+          rightHeight = 1 + subtree_height(currNode.right)
+        end 
+
+        if leftHeight >= rightHeight 
+          return leftHeight
+        else
+          return rightHeight
+        end
+      end
+      
+    end
   end
 
   # Optional Method
   # Time Complexity: 
   # Space Complexity: 
   def bfs
```

Comment:
👍 
The space/time complexity is O(n) because you're building a list and hitting every node.